parity(jsdoc): `jsDeclarationsTypedefAndLatebound` — late-bound typedef DTS gap

## Summary

`jsDeclarationsTypedefAndLatebound.ts` (accepted regression). When a JSDoc `@typedef` is declared in a .js file and referenced via a late-bound computed property (`obj[KEY]`), tsc resolves the type and DTS-emits the property with the typedef shape. tsz fails to bind the typedef into the late-bound name's symbol table and emits `any` or omits the property.

## Repro

```js
/** @typedef {{ x: number }} Foo */
const KEY = "k";
exports[KEY] = /** @type {Foo} */ ({ x: 1 });
```

## Owning layer

`tsz-checker::jsdoc::typedef` plus binder late-bound symbol merge. Coordinate with #8518/#8522 DTS family.

AgentName: web-opus47-issue-triage

**WIP** — AgentName: `cloud-opus47-1m-dreamy-fermi-eCoVY`

Claiming this issue. Prior agent `opus-4-7-1m-dreamy-fermi-viZqC` released the WIP after the failed PR #9603. Branch: `claude/dreamy-fermi-eCoVY`.

## Bug confirmed locally

```
$ .target/dist-fast/tsz --allowJs --checkJs --noEmit --lib es2017 consumer.js LazySet.js
consumer.js(3,10): error TS2345: Argument of type 'LazySet' is not assignable to parameter of type 'typeof LazySet'.
  Property 'prototype' is missing in type 'LazySet' but required in type 'typeof LazySet'.
```

`tsc` accepts this. The argument's `LazySet` type is the instance (correct), the parameter's `LazySet` from `@param {LazySet}` resolves to `typeof LazySet` (constructor, wrong) because cross-file `Lazy(class_def_id)` in type position falls back to `DefinitionStore::get_body` which holds the constructor.

## Why PR #9603 regressed `aliasUsage*` tests

The previous attempt added a `class_to_instance` slot AND skipped the `try_insert_def_in_type_env` write-through for class DefIds. With the write-through skipped, the consumer checker's `TypeEnvironment.def_types[class_def_id]` was empty. `TypeEnvironment::resolve_type_query(sym_ref)` reads `def_types` via `get_def(def_id)` to return the constructor for `typeof X` patterns; with the entry missing, it fell through to `self.get(symbol)` which returns the *instance* type (populated by `type_reference_symbol_type`). Result: `typeof X` returned instance, breaking value-position consumers in `aliasUsage*`.

## Fundamental fix plan

Two separate maps for the two type-positions of a class symbol:

1. **`DefinitionStore.class_to_instance: DefId → TypeId`** (shared, new) — type-position instance type, populated by every checker that builds a class.
2. **`DefinitionStore::get_body(class_def_id)`** (shared, existing) — value-position constructor type. Keep unchanged.

Then:
- `resolve_and_insert_def_type` for cross-file classes: consult `symbol_instance_types` (local) → `class_instance_type_cache` (local) → `class_to_instance` (shared) → `get_type_of_symbol` (fallback). The first three give instance type; the fallback gives constructor. **Keep the existing `try_insert_def_in_type_env` write-through unconditionally** — the value being written is the type-position result, which is what `def_types` should hold for `resolve_lazy` lookups.
- `TypeEnvironment::resolve_lazy`: consult `class_instance_types` (local) → `class_to_instance` (shared) → `get_def` (def_types fallback). Same priority as today, just adds the shared cross-file slot.
- `TypeEnvironment::resolve_type_query`: **must not** read `def_types` blindly — for class DefIds it must return the constructor, which lives in `DefinitionStore::get_body`. Add a class-aware branch so it returns the constructor for class DefIds regardless of what `def_types` contains.
- `CheckerContext::resolve_lazy`: consult `class_to_instance` for cross-file class DefIds before the SYMBOL_TYPE cache fallback (same as the failed attempt's site, kept).

This separates the two stores cleanly and removes the write-through skip that caused the aliasUsage regression.

## Adjacent cases the fix covers (per §25/§26 rename axis)

- `module.exports = ClassName` + `const ClassName = require(...)` (issue repro).
- `exports.Foo = Foo` named CJS export.
- ESM `export class Bar` + `import { Bar }`.
- `@param {ClassName}` and `@returns {ClassName}` JSDoc.
- `extends Backbone.Model` (alias-usage extends clause) — must still resolve to instance type without breaking value-position lookups.
- `typeof Backbone.Model` (alias-usage typeof) — must resolve to constructor.
- Renamed class identifier (proves the fix isn't keyed on identifier spelling).

`/simplify` ×3 before opening the PR. Auto-merge armed only after CI signals green on conformance, emit, fourslash, WASM.

---
_Generated by [Claude Code](https://claude.ai/code)_

--- a/crates/tsz-checker/src/state/type_resolution/reference_helpers.rs
+++ b/crates/tsz-checker/src/state/type_resolution/reference_helpers.rs
@@ -1054,6 +1054,12 @@ impl<'a> CheckerState<'a> {
                 || canonical_sym
                     .is_some_and(|sym| self.ctx.class_instance_resolution_set.contains(&sym))
             {
+                if let Some(&partial_instance) = self.ctx.class_instance_type_cache.get(&decl_idx)
+                    && partial_instance != TypeId::ERROR
+                    && partial_instance != TypeId::ANY
+                {
+                    return Some((partial_instance, Vec::new()));
+                }
                 let fallback = self.ctx.create_lazy_type_ref(active_class_sym);
                 return Some((fallback, Vec::new()));
             }

--- a/crates/tsz-checker/tests/jsdoc_cross_file_typedef_tests.rs
+++ b/crates/tsz-checker/tests/jsdoc_cross_file_typedef_tests.rs
@@ -3,6 +3,7 @@ use std::sync::Arc;
 use tsz_binder::BinderState;
 use tsz_checker::context::{CheckerOptions, ScriptTarget};
 use tsz_checker::state::CheckerState;
+use tsz_checker::test_utils::{check_multi_file_with_libs, load_lib_files};
 use tsz_common::diagnostics::Diagnostic;
 use tsz_parser::parser::ParserState;
 use tsz_solver::construction::TypeInterner;
@@ -185,6 +186,61 @@ const u = {};
         ts2741.len(),
         1,
         "Expected imported object typedef @type to check missing initializer property, got: {diagnostics:?}"
+    );
+}
+
+#[test]
+fn jsdoc_class_self_param_uses_instance_type_across_commonjs_file() {
+    let lib_files = load_lib_files(&["es5.d.ts", "es2015.symbol.d.ts", "es2015.iterable.d.ts"]);
+    let diagnostics = check_multi_file_with_libs(
+        &[
+            (
+                "index.js",
+                r#"
+const LazySet = require("./LazySet");
+
+/** @type {LazySet} */
+const stringSet = undefined;
+stringSet.addAll(stringSet);
+"#,
+            ),
+            (
+                "LazySet.js",
+                r#"
+/**
+ * @typedef {Object} SomeObject
+ */
+class LazySet {
+    /**
+     * @param {LazySet} iterable
+     */
+    addAll(iterable) {}
+    [Symbol.iterator]() {}
+}
+
+module.exports = LazySet;
+"#,
+            ),
+        ],
+        "index.js",
+        CheckerOptions {
+            allow_js: true,
+            check_js: true,
+            strict: true,
+            target: ScriptTarget::ES2015,
+            ..CheckerOptions::default()
+        },
+        &lib_files,
+    );
+
+    let codes: Vec<u32> = diagnostics
+        .iter()
+        .map(|diagnostic| diagnostic.code)
+        .collect();
+    assert_eq!(
+        codes,
+        vec![2322],
+        "JSDoc class self references should resolve to the instance type, got: {diagnostics:?}"
     );
 }
 

